### PR TITLE
[RooFit] Improve doxygen

### PR DESCRIPTION
**Pull request by hageboeck**
Re-format important roofit documentation to make doxygen understand it.

Also make histfactory classes show up in doxygen.

### CODE REVIEW

File: roofit/histfactory/inc/RooStats/HistFactory/Channel.h [hunk 1]
```diff
@@ -33,41 +33,41 @@ class Channel  {
   Channel(const Channel& other);
   Channel(std::string Name, std::string InputFile="");
 
-  // set name of channel
+  //// set name of channel
```
Comment (dpiparo) [inline roofit/histfactory/inc/RooStats/HistFactory/Channel.h]: why 4 '/' ? :)

File: roofit/histfactory/src/Data.cxx [hunk 2]
```diff
@@ -9,14 +9,10 @@
...
 ////////////////////////////////////////////////////////////////////////////////
-
-/*
-BEGIN_HTML
-<p>
-</p>
-END_HTML
+/** \class RooStats::HistFactory::Data
+ *  \ingroup HistFactory 
+ * TODO Add documentation.
```
Comment (dpiparo) [inline roofit/histfactory/src/Data.cxx]: Why this todo? perhaps it's better to drop it...

File: roofit/histfactory/src/ParamHistFunc.cxx [hunk 3]
```diff
@@ -10,13 +10,30 @@
...
-</p>
-END_HTML
-*/
-//
+/** \class ParamHistFunc
+ * \ingroup HistFactory
+ *   A class which maps the current values of a RooRealVar
+ *  (or a set of RooRealVars) to one of a number of RooRealVars:
+
+ *  ParamHistFunc: {val1, val2, ...} -> {gamma (RooRealVar)}
```
Comment (dpiparo) [inline roofit/histfactory/src/ParamHistFunc.cxx]: maybe this looks better enclosed in reverse quotes?

File: roofit/histfactory/src/ParamHistFunc.cxx [hunk 4]
```diff
@@ -10,13 +10,30 @@
...
+
+ *  The 'createParamSet' is an easy way to create these
+ *  parameters from a set of observables. They are
+ *  stored using the "TH1" ordering convention (as compared
+ *  to the RooDataHist convention, which is used internally
+ *  and one must map between the two).
+
+ *  All indices include '0'
+
+ *  gamma_i_j = paramSet[ size(i)*j + i ]
```
Comment (dpiparo) [inline roofit/histfactory/src/ParamHistFunc.cxx]: Does this become simple text or a veritable formula/formula like block?

File: roofit/histfactory/src/ParamHistFunc.cxx [hunk 5]
```diff
@@ -10,13 +10,30 @@
...
+
+ *  ie assuming the dimensions are 5*5:
+ *  gamma_2_1 = paramSet[ 5*1 + 2 ] = paramSet[7]
```
Comment (dpiparo) [inline roofit/histfactory/src/ParamHistFunc.cxx]: say comment about formula block above.

File: roofit/histfactory/src/ParamHistFunc.cxx [hunk 6]
```diff
@@ -93,7 +83,17 @@ ParamHistFunc::ParamHistFunc() : _numBins(0)
...
 ////////////////////////////////////////////////////////////////////////////////
-
+/// Create a function which returns binewise-values
+/// This class contains N RooRealVar's, one for each
+/// bin from the given RooRealVar.
+///
+/// The value of the function in the ith bin is 
+/// given by:
+///
+/// F(i) = gamma_i * nominal(i)
```
Comment (dpiparo) [inline roofit/histfactory/src/ParamHistFunc.cxx]: formula?

File: roofit/histfactory/src/ParamHistFunc.cxx [hunk 7]
```diff
@@ -136,7 +125,17 @@ ParamHistFunc::ParamHistFunc(const char* name, const char* title,
...
```
Comment (dpiparo) [inline roofit/histfactory/src/ParamHistFunc.cxx]: formula..

File: roofit/histfactory/src/ParamHistFunc.cxx [hunk 8]
```diff
@@ -227,31 +214,27 @@ ParamHistFunc::~ParamHistFunc()
...
 
 ////////////////////////////////////////////////////////////////////////////////
-
+/// Get the index of the gamma parameter associated
+/// with the current bin
+/// This number is the "RooDataSet" style index
+/// and it must be because it uses the RooDataSet method directly
+/// This is intended to be fed into the getParameter(Int_t) method:
+///
+/// RooRealVar currentParam = getParameter( getCurrentBin() );
```
Comment (dpiparo) [inline roofit/histfactory/src/ParamHistFunc.cxx]: codeblock?
```{.cpp}
 RooRealVar currentParam = getParameter( getCurrentBin() );
```

File: roofit/histfactory/src/ParamHistFunc.cxx [hunk 9]
```diff
@@ -454,22 +437,22 @@ RooArgList ParamHistFunc::createParamSet(RooWorkspace& w, const std::string& Pre
...
+/// parameters which represent the
+/// height of the histogram bins.
+/// The list 'vars' represents the 
+/// observables (corresponding to histogram bins)
+/// that these newly created parameters will 
+/// be mapped to. (ie, we create one parameter
+/// per observable in vars and per bin in each observable)
+
+/// Store them in a list using:
+/// _paramSet.add( createParamSet() );
```
Comment (dpiparo) [inline roofit/histfactory/src/ParamHistFunc.cxx]: code block?

File: roofit/roofitcore/src/RooAbsData.cxx [hunk 10]
```diff
@@ -346,15 +346,16 @@ void RooAbsData::setDirtyProp(Bool_t flag)
 /// Create a reduced copy of this dataset. The caller takes ownership of the returned dataset
 ///
 /// The following optional named arguments are accepted
-///
-///  - `SelectVars(const RooArgSet& vars)` Only retain the listed observables in the output dataset
-///  - `Cut(const char* expression)` Only retain event surviving the given cut expression
-///  - `Cut(const RooFormulaVar& expr)` Only retain event surviving the given cut formula
-///  - `CutRange(const char* name)` Only retain events inside range with given name. Multiple CutRange
+/// <table>
+/// <tr><td> SelectVars(const RooArgSet& vars)  <td> Only retain the listed observables in the output dataset
```
Comment (dpiparo) [inline roofit/roofitcore/src/RooAbsData.cxx]: maybe this first column can be enclosed between reverse quotes?

File: roofit/roofitcore/src/RooAbsData.cxx [hunk 11]
```diff
@@ -683,7 +624,30 @@ TH1 *RooAbsData::createHistogram(const char *name, const RooAbsRealLValue& xvar,
 }
 
 ////////////////////////////////////////////////////////////////////////////////
-/// Internal method that implements histogram filling
+///
+/// This function accepts the following arguments
+///
+/// \param[in] name Name of the ROOT histogram
+/// \param[in] xvar Observable to be mapped on x axis of ROOT histogram
```
Comment (dpiparo) [inline roofit/roofitcore/src/RooAbsData.cxx]: WOW!
perhaps also the return value description?

File: roofit/roofitcore/src/RooAbsData.cxx [hunk 12]
```diff
@@ -1150,24 +1117,30 @@ RooRealVar* RooAbsData::rmsVar(RooRealVar &var, const char* cutSpec, const char*
...
-///  - `Layout(Double_t xmin, Double_t xmax, Double_t ymax)` Specify relative position of left,right side of box and top of box. Position of
+/// <tr><td> Format(const char* what,...)  <td> Parameter formatting options.
+///   <table>
+///   <tr><td> const char* what <td> Controls what is shown:
+///     - "N" adds name
+///     - "E" adds error
+///     - "A" shows asymmetric error
+///     - "U" shows unit
+///     - "H" hides the value
+///   <tr><td> FixedPrecision(int n)  <td> Controls precision, set fixed number of digits
```
Comment (dpiparo) [inline roofit/roofitcore/src/RooAbsData.cxx]: again, same comment: perhaps code can be properly hightlighted?

File: roofit/roofitcore/src/RooAbsPdf.cxx [hunk 13]
```diff
@@ -3061,14 +2987,13 @@ RooAbsReal* RooAbsPdf::createCdf(const RooArgSet& iset, const RooArgSet& nset)
...
 /// The following named arguments are accepted
-///
-/// SupNormSet(const RooArgSet&)         -- Observables over which should be normalized _in_addition_ to the
-///                                         integration observables
-/// ScanNumCdf()                         -- Apply scanning technique if cdf integral involves numeric integration [ default ] 
-/// ScanAllCdf()                         -- Always apply scanning technique 
-/// ScanNoCdf()                          -- Never apply scanning technique                  
-/// ScanParameters(Int_t nbins,          -- Parameters for scanning technique of making CDF: number
-///                Int_t intOrder)          of sampled bins and order of interpolation applied on numeric cdf
+/// | Type of CmdArg    |    Effect on CDF
```
Comment (dpiparo) [inline roofit/roofitcore/src/RooAbsPdf.cxx]: does this render as a table? WOW

File: roofit/roofitcore/src/RooAbsPdf.cxx [hunk 14]
```diff
@@ -2213,13 +2131,14 @@ RooDataHist *RooAbsPdf::generateBinned(const RooArgSet& whatVars, Double_t nEven
 /// dataset.
 ///
 /// The following named arguments are supported
+/// | Type of CmdArg       | Effect on generation
+/// |-------------------------|-----------------------
+/// | Name(const char* name)  | Name of the output dataset
```
Comment (dpiparo) [inline roofit/roofitcore/src/RooAbsPdf.cxx]: same comment as usual: perhaps c++ code can be highlighted?

File: roofit/histfactory/inc/RooStats/HistFactory/Channel.h (hunk 1 above)
Comment (hageboeck) [reply roofit/histfactory/inc/RooStats/HistFactory/Channel.h]: Typo ...

File: roofit/histfactory/src/ParamHistFunc.cxx (hunk 3 above)
Comment (hageboeck) [reply roofit/histfactory/src/ParamHistFunc.cxx]: Yeah ...
A comment I didn't touch at all. Just pulled it to the right place ...

Going the extra mile for you now ...

File: roofit/roofitcore/src/RooAbsData.cxx (hunk 10 above)
Comment (hageboeck) [reply roofit/roofitcore/src/RooAbsData.cxx]: Let's see if my skills with sed can save me tons of work ...

Comment (hageboeck) [reply roofit/roofitcore/src/RooAbsData.cxx]: Ok, call me regex master!

**Comment by hageboeck** [comment]
@phsft-bot build just on windows10/vc15

**Comment by bellenot** [comment]
@phsft-bot build just on windows10/vc15

**Comment by bellenot** [comment]
@phsft-bot build just on windows10/vc15

**Comment by bellenot** [comment]
@phsft-bot build just on windows10/vc15

**Comment by hageboeck** [comment]
@phsft-bot build just on windows10/vc15

**Comment by hageboeck** [comment]
@phsft-bot build just on windows10/vc15

**Pull request merged**